File: mymath/collision.go
package mymath

import (
	"math"
)

// Overlap from rect to circle
func CircleRectOverlap(c Circle, r Rect) (bool, Vec) {
	closestLine := r.ClosestSide(c.Pos)
	closestPoint := closestLine.ClosestPoint(c.Pos)

	centreInside := r.ContainsPoint(c.Pos)

	var overlapAxis Vec
	if centreInside {
		overlapAxis = c.Pos.Sub(closestPoint)
	} else {
		overlapAxis = closestPoint.Sub(c.Pos)
	}

	dist := overlapAxis.Length()
	if dist >= c.Radius && !centreInside {
		return false, Vec{}
	}

	// Make sepAxis unit vector
	if dist < 1e-6 {
		rectMid := r.Pos.Add(r.Size.Scale(0.5))
		overlapAxis = rectMid.Sub(closestPoint).Normalize()
	} else {
		overlapAxis = overlapAxis.Scale(1 / dist)
	}
	return true, overlapAxis.Scale(c.Radius - dist)
}

// lines: ccw lines of convex polygon
func CircleTriangleOverlap(circle Circle, t0 Vec, t1 Vec, t2 Vec) (bool, Vec) {
	// Triangle points and normals
	//        t2
	//        /\
	// n2 <- /  \ -> n1
	//      /    \
	//  t0 /______\ t1
	//         |
	//        n0

	// vectors from tri vertices to circle centre
	u0 := circle.Pos.Sub(t0)
	u1 := circle.Pos.Sub(t1)
	u2 := circle.Pos.Sub(t2)

	n0 := t1.Sub(t0).Normalize()
	n0.X, n0.Y = n0.Y, -n0.X

	u0DotNormal := u0.Dot(n0)
	if u0DotNormal > 0 { // below triangle
		return circleTriangleSideOverlap(circle, u0, u1, t0, t1, n0, u0DotNormal)
	}

	n1 := t2.Sub(t1).Normalize()
	n1.X, n1.Y = n1.Y, -n1.X
	u1DotNormal := u1.Dot(n1)
	if u1DotNormal > 0 { // below triangle
		return circleTriangleSideOverlap(circle, u1, u2, t1, t2, n1, u1DotNormal)
	}

	n2 := t0.Sub(t2).Normalize()
	n2.X, n2.Y = n2.Y, -n2.X
	u2DotNormal := u2.Dot(n2)
	if u2DotNormal > 0 { // below triangle
		return circleTriangleSideOverlap(circle, u2, u0, t2, t0, n2, u2DotNormal)
	}

	// circle centre withing triangle, overlap is from nearest side
	if u0DotNormal > u1DotNormal {
		if u0DotNormal > u2DotNormal {
			return true, n0.Scale(u0DotNormal - circle.Radius)
		}
		return true, n2.Scale(u2DotNormal - circle.Radius)
	}
	if u1DotNormal > u2DotNormal {
		return true, n1.Scale(u1DotNormal - circle.Radius)
	}
	return true, n2.Scale(u2DotNormal - circle.Radius)
}

func circleTriangleSideOverlap(circle Circle, u0 Vec, u1 Vec, t0 Vec, t1 Vec, n0 Vec, u0DotNormal float64) (bool, Vec) {
	if u0DotNormal >= circle.Radius {
		return false, Vec{}
	}

	l0 := t1.Sub(t0)
	l0Dotu0 := l0.Dot(u0)
	if l0Dotu0 < 0 { // closest to left point
		u0Len := u0.Length()
		if u0Len >= circle.Radius {
			return false, Vec{}
		}
		return true, u0.Scale((u0Len - circle.Radius) / u0Len)
	}

	if l0Dotu0 > l0.SqrLength() { // closest to right point
		u1Len := u1.Length()
		if u1Len >= circle.Radius {
			return false, Vec{}
		}
		return true, u1.Scale((u1Len - circle.Radius) / u1Len)
	}

	// perpendicular to bottom line
	if u0DotNormal >= circle.Radius {
		return false, Vec{}
	}
	return true, n0.Scale(u0DotNormal - circle.Radius)
}

// Returns intersection closest to start of line. If the start of the line is within the circle, returns this point.
func LaserCircleIntersect(l Line, circle Circle) (bool, Vec) {
	u := l.Start.Sub(circle.Pos)
	v := l.End.Sub(l.Start)
	a := v.Dot(v)
	b := 2 * u.Dot(v)
	c := u.Dot(u) - circle.Radius*circle.Radius
	discriminant := b*b - 4*a*c
	if discriminant < 0 {
		return false, u
	}
	sqrt := math.Sqrt(discriminant)
	t1 := (-b - sqrt) / (2 * a)
	t2 := (-b + sqrt) / (2 * a)
	if t2 <= 0 || t1 >= 1 {
		return false, u
	}
	return true, l.Start.Add(v.Scale(t1))
}

type Tiling struct {
	Left   bool
	Right  bool
	Top    bool
	Bottom bool
}

func LaserRectIntersect(l Line, r Rect) (bool, Vec, Vec) {
	// Avoid case of laser beggining slightly inside shape (e.g. after a bounce)
	if r.ContainsPoint(l.Start) {
		return false, Vec{}, Vec{}
	}

	lineDir := l.End.Sub(l.Start)
	var intersectsVertical bool
	var intersectionVertical Vec
	var normalVertical Vec
	if lineDir.X > 0 {
		intersectsVertical, intersectionVertical = l.Intersection(r.LeftLine())
		normalVertical = Vec{-1, 0}
	} else if lineDir.X < 0 {
		intersectsVertical, intersectionVertical = l.Intersection(r.RightLine())
		normalVertical = Vec{1, 0}
	}

	var intersectsHorizontal bool
	var intersectionHorizontal Vec
	var normalHorizontal Vec
	if lineDir.Y > 0 {
		intersectsHorizontal, intersectionHorizontal = l.Intersection(r.BottomLine())
		normalHorizontal = Vec{0, -1}
	} else if lineDir.Y < 0 {
		intersectsHorizontal, intersectionHorizontal = l.Intersection(r.TopLine())
		normalHorizontal = Vec{0, 1}
	}

	if intersectsVertical && intersectsHorizontal {
		distVertical := intersectionVertical.DistanceTo(l.Start)
		distHorizontal := intersectionHorizontal.DistanceTo(l.Start)
		if distVertical < distHorizontal {
			return true, intersectionVertical, normalVertical
		}
		return true, intersectionHorizontal, normalHorizontal
	}
	if intersectsVertical {
		return true, intersectionVertical, normalVertical
	}
	if intersectsHorizontal {
		return true, intersectionHorizontal, normalHorizontal
	}
	return false, intersectionVertical, normalVertical
}

func LaserTriangleIntersect(line Line, t0 Vec, t1 Vec, t2 Vec) (bool, Vec, Vec) {
	closestDist := -1.0
	var closestIntersect Vec
	var closestNormal Vec

	dist, intersect, normal := laserTriangleSideIntersect(line, t0, t1)
	if dist >= 0 && (closestDist < 0 || dist < closestDist) {
		closestDist = dist
		closestIntersect = intersect
		closestNormal = normal
	}

	dist, intersect, normal = laserTriangleSideIntersect(line, t1, t2)
	if dist >= 0 && (closestDist < 0 || dist < closestDist) {
		closestDist = dist
		closestIntersect = intersect
		closestNormal = normal
	}

	dist, intersect, normal = laserTriangleSideIntersect(line, t2, t0)
	if dist >= 0 && (closestDist < 0 || dist < closestDist) {
		closestDist = dist
		closestIntersect = intersect
		closestNormal = normal
	}

	return closestDist >= 0, closestIntersect, closestNormal
}

func laserTriangleSideIntersect(line Line, t0 Vec, t1 Vec) (float64, Vec, Vec) {
	normal := clockWiseNormal(t1.Sub(t0))
	if line.Start.Sub(t0).Dot(normal) > 0 { // Avoid case of laser beggining slightly inside shape (e.g. after a bounce)
		side := Line{t0, t1}
		intersected, intersect := line.Intersection(side)
		if intersected {
			dist := intersect.DistanceTo(line.Start)
			return dist, intersect, normal
		}
	}
	return -1.0, normal, normal
}

func clockWiseNormal(dir Vec) Vec {
	normal := dir.Normalize()
	normal.X, normal.Y = normal.Y, -normal.X
	return normal
}

File: www/collision/collision.js
import {Vec, Line} from "../math.js";

// Overlap from rect to circle
function circleRectOverlap(c, r) {
	const closestLine = r.closestSide(c.pos);
	const closestPoint = closestLine.closestPoint(c.pos);

	const centreInside = r.containsPoint(c.pos);

	let overlapAxis;
	if (centreInside) {
		overlapAxis = c.pos.sub(closestPoint);
	} else {
		overlapAxis = closestPoint.sub(c.pos);
	}

	let dist = overlapAxis.length();
	if (dist >= c.radius && !centreInside) {
		return null
	}

	// If centre of circle on side of rect, need to make sure axis is non-zero
	if (dist < 1e-6) {
		const rectMid = r.pos.add(r.size.scale(0.5));
		overlapAxis = rectMid.sub(closestPoint).normalize();
	} else {
		overlapAxis = overlapAxis.scale(1/dist);
	}

	return overlapAxis.scale(c.radius - dist);
}

// lines: ccw lines of convex polygon
function circleTriangleOverlap(circle, t0, t1, t2) {
	// Triangle points and normals
	//        t2
	//        /\
	// n2 <- /  \ -> n1
	//      /    \
	//  t0 /______\ t1
	//         |
	//        n0

	const n0 = t1.sub(t0).normalize();
	let tmpX = n0.x;
	n0.x = n0.y;
	n0.y = -tmpX;
	const u0 = circle.pos.sub(t0);
	const u1 = circle.pos.sub(t1);
	const u2 = circle.pos.sub(t2);

	const u0DotNormal = u0.dot(n0);
	if (u0DotNormal > 0) { // below triangle
		return _circleTriangleSideOverlap(circle, u0, u1, t0, t1, n0, u0DotNormal);
	}

	const n1 = t2.sub(t1).normalize();
	tmpX = n1.x;
	n1.x = n1.y;
	n1.y = -tmpX;
	const u1DotNormal = u1.dot(n1);
	if (u1DotNormal > 0) { // below triangle
		return _circleTriangleSideOverlap(circle, u1, u2, t1, t2, n1, u1DotNormal);
	}

	const n2 = t0.sub(t2).normalize();
	tmpX = n2.x;
	n2.x = n2.y;
	n2.y = -tmpX;
	const u2DotNormal = u2.dot(n2);
	if (u2DotNormal > 0) { // below triangle
		return _circleTriangleSideOverlap(circle, u2, u0, t2, t0, n2, u2DotNormal);
	}

	// circle centre withing triangle, overlap is from nearest side
	if (u0DotNormal > u1DotNormal) {
		if (u0DotNormal > u2DotNormal) {
			return n0.scale(u0DotNormal - circle.radius);
		}
		return n2.scale(u2DotNormal - circle.radius);
	}
	if (u1DotNormal > u2DotNormal) {
		return n1.scale(u1DotNormal - circle.radius);
	}
	return n2.scale(u2DotNormal - circle.radius);
}

function _circleTriangleSideOverlap(circle, u0, u1, t0, t1, n0, u0DotNormal) {
	if (u0DotNormal >= circle.radius) {
		return null;
	}

	const l0 = t1.sub(t0);
	const l0Dotu0 = l0.dot(u0);
	if (l0Dotu0 < 0) { // closest to left point
		const u0Len = u0.length();
		if (u0Len >= circle.radius) {
			return null;
		}
		return u0.scale((u0Len - circle.radius)/u0Len);
	}

	if (l0Dotu0 > l0.sqrLength()) { // closest to right point
		const u1Len = u1.length();
		if (u1Len >= circle.radius) {
			return null;
		}
		return u1.scale((u1Len - circle.radius)/u1Len);
	}

	// perpendicular to bottom line
	if (u0DotNormal >= circle.radius) {
		return null;
	}
	return n0.scale(u0DotNormal - circle.radius);
}

// Returns intersection closest to start of line. If the start of the line is within the circle, returns this point.
function laserCircleIntersect(l, circle) {
	let u = l.start.sub(circle.pos);
	let v = l.end.sub(l.start);
	let a = v.dot(v);
	let b = 2 * u.dot(v);
	let c = u.dot(u) - circle.radius*circle.radius;
	let discriminant = b*b - 4*a*c;
	if (discriminant < 0) {
		return null;
	}

	let sqrt = Math.sqrt(discriminant);
	let t1 = (-b - sqrt) / (2 * a);
	let t2 = (-b + sqrt) / (2 * a);
	if (t2 <= 0 || t1 >= 1) {
		return null;
	}
	return l.start.add(v.scale(t1));
}

// Return intersection closest to start of line.
function laserRectIntersect(line, rect) {
	// Avoid case of laser beggining slightly inside shape (e.g. after a bounce)
	if (rect.containsPoint(line.start)) {
		return [null, null];
	}

	const lineDir = line.end.sub(line.start);
	let intersectionVertical;
	let normalVertical;
	if (lineDir.x > 0) {
		intersectionVertical = line.intersection(rect.leftLine());
		normalVertical = new Vec(-1, 0);
	} else if (lineDir.x < 0) {
		intersectionVertical = line.intersection(rect.rightLine());
		normalVertical = new Vec(1, 0);
	}

	let intersectionHorizontal;
	let normalHorizontal;
	if (lineDir.y > 0) {
		intersectionHorizontal = line.intersection(rect.bottomLine());
		normalHorizontal = new Vec(0, -1);
	} else if (lineDir.y < 0) {
		intersectionHorizontal = line.intersection(rect.topLine());
		normalHorizontal = new Vec(0, 1);
	}

	if (intersectionVertical && intersectionHorizontal) {
		let distVertical = intersectionVertical.distanceTo(line.start);
		let distHorizontal = intersectionHorizontal.distanceTo(line.start);
		if (distVertical < distHorizontal) {
			return [intersectionVertical, normalVertical];
		} else {
			return [intersectionHorizontal, normalHorizontal];
		}
	}
	if (intersectionVertical) {
		return [intersectionVertical, normalVertical];
	}
	if (intersectionHorizontal) {
		return [intersectionHorizontal, normalHorizontal];
	}
	return [null, null];
}

// Return intersection closest to start of line.
function laserTriangleIntersect(line, t0, t1, t2) {
	let closestDist = -1.0;
	let closestIntersect = null;
	let closestNormal = null;

	let [dist, intersect, normal] = _laserTriangleSideIntersect(line, t0, t1)
	if (dist >= 0 && (closestDist < 0 || dist < closestDist)) {
		closestDist = dist
		closestIntersect = intersect
		closestNormal = normal
	}

	[dist, intersect, normal] = _laserTriangleSideIntersect(line, t1, t2)
	if (dist >= 0 && (closestDist < 0 || dist < closestDist)) {
		closestDist = dist
		closestIntersect = intersect
		closestNormal = normal
	}

	[dist, intersect, normal] = _laserTriangleSideIntersect(line, t2, t0)
	if (dist >= 0 && (closestDist < 0 || dist < closestDist)) {
		closestDist = dist
		closestIntersect = intersect
		closestNormal = normal
	}
	return [closestIntersect, closestNormal];
}

function _laserTriangleSideIntersect(line, t0, t1) {
	const normal = _clockWiseNormal(t1.sub(t0));
	if (line.start.sub(t0).dot(normal) > 0) { // Avoid case of laser beggining slightly inside shape (e.g. after a bounce)
		const side = new Line(t0, t1);
		const intersect = line.intersection(side);
		if (intersect !== null) {
			let dist = intersect.distanceTo(line.start)
			return [dist, intersect, normal];
		}
	}

	return [-1.0, null, null];
}

function _clockWiseNormal(dir) {
	const normal = dir.normalize();
	const tmpX = normal.x;
	normal.x = normal.y;
	normal.y = -tmpX;
	return normal;
}

export {circleRectOverlap, circleTriangleOverlap, laserCircleIntersect, laserRectIntersect, laserTriangleIntersect};